src/opsinputs/opsinputs_fill_mod.F90

```diff
@@ -1067,41 +1072,58 @@ subroutine opsinputs_fill_fillreal2d_norecords( &
 !take into account offsetting of 2nd dimension if required
 !designed to be used to pack where multiple satellite instruments expected
 !e.g. HIRS in ATOVS stream
-offset = 0
+
 numchans = size(JediVarNamesWithChannels)
-if (present(OffsetChans)) then
-  offset = OffsetChans % channel_offset
-  if (OffsetChans % size_of_varobs_array > 0) &
-    numchans = OffsetChans % size_of_varobs_array
-end if
 
-!Setup for channels needing to match array index
-localUseActualChans = .false.
-if (present(useActualChans)) then
-    localUseActualChans = useActualChans
+if (present(sizeVarObs) )then
+  if (sizeVarObs > 0) then
+    numchans = sizeVarObs
+  end if
 end if
 
 if (obsspace_has(ObsSpace, JediVarGroup, JediVarNamesWithChannels(1))) then
   ! Allocate OPS data structures
   call Ops_Alloc(Hdr, OpsVarName, NumObs, Real2, &
                  num_levels = int(numchans, kind=integer64))
   do iChannel = 1, size(JediVarNamesWithChannels)
-    ! Retrieve data from JEDI
-    call obsspace_get_db(ObsSpace, JediVarGroup, JediVarNamesWithChannels(iChannel), VarValue)
-
-    ! Fill the OPS data structures
-    if (localUseActualChans) then
-      where (VarValue /= MissingDouble)
-        Real2(:, Channels(iChannel)) = VarValue
-      end where
+      call obsspace_get_db(ObsSpace, JediVarGroup, JediVarNamesWithChannels(iChannel), VarValue)
+
+    if (present(varChannels)) then
+      if (size(varChannels) > 0) then
+        if (iChannel <= size(varChannels)) then
+          if (compressVarChannels) then
+            offsetsize = abs(varChannels(1) - channels(1))
+            where (VarValue /= MissingDouble)
+              Real2(:, iChannel+offsetsize) = VarValue
+            end where
+          else
+            where (VarValue /= MissingDouble)
+              Real2(:, varChannels(iChannel)) = VarValue
+            end where
+          end if
+        end if
+      end if
     else
-     ! Fill the OPS data structures
-      where (VarValue /= MissingDouble)
-        Real2(:, iChannel+offset) = VarValue
-      end where
+      if ((present(sizeVarObs)) .and. (sizeVarObs > size(channels))) then
```

Comment:
These should be on separate clauses.  There is no guarantee in Fortran that the first clause is evaluated before the second.

Reply:
Have now changed